File: lib/plotting/stages/isoform_calling.py
"""Isoform calling stage — long-read RNA-seq taxonomy. Renders the
transcript structure plot for the top isoform models and a sashimi
plot of splice junctions in a representative locus.

Plan reference: §S13 (long-read-rnaseq taxonomy stage `isoform_calling`).

Manifest contract:
- ``isoform_table``: TSV with `transcript`, `exon_starts`, `exon_ends`,
  optional `strand`. exon_starts/ends are comma-separated lists of ints.
- ``junction_table``: TSV with `junction` ("start-end" string), `count`.
"""

from __future__ import annotations

from pathlib import Path
from typing import List, Optional

from ..core import (
    FigureContext,
    isoform_structure,
    register_figure,
    sashimi,
    stage_registry,
)
from ._shared import load_tsv_columns, manifest_path, open_text

FIGURES = stage_registry("isoform_calling")


def _parse_int_list(s: str) -> List[int]:
    if not s:
        return []
    return [int(x) for x in s.split(",") if x.strip()]


def _load_isoforms(p: Path) -> Optional[dict]:
    """Custom loader for isoform tables — load_tsv_columns can't handle
    the comma-list cells, so we parse them inline.
    """
    transcripts: List[str] = []
    exon_starts: List[List[int]] = []
    exon_ends: List[List[int]] = []
    strand: List[str] = []
    try:
        with open_text(p) as f:
            header = f.readline().rstrip("\n").split("\t")
            t_idx = next(
                (i for i, h in enumerate(header)
                 if h.lower() in ("transcript", "transcript_id", "tx")),
                None,
            )
            s_idx = next(
                (i for i, h in enumerate(header)
                 if h.lower() in ("exon_starts", "starts", "blockstarts")),
                None,
            )
            e_idx = next(
                (i for i, h in enumerate(header)
                 if h.lower() in ("exon_ends", "ends", "blockends")),
                None,
            )
            sd_idx = next(
                (i for i, h in enumerate(header) if h.lower() in ("strand",)),
                None,
            )
            if t_idx is None or s_idx is None or e_idx is None:
                return None
            for line in f:
                parts = line.rstrip("\n").split("\t")
                if len(parts) <= max(t_idx, s_idx, e_idx):
                    continue
                transcripts.append(parts[t_idx])
                exon_starts.append(_parse_int_list(parts[s_idx]))
                exon_ends.append(_parse_int_list(parts[e_idx]))
                strand.append(parts[sd_idx] if sd_idx is not None and sd_idx < len(parts) else "+")
    except OSError:
        return None
    if not transcripts:
        return None
    return {
        "transcript": transcripts,
        "exon_starts": exon_starts,
        "exon_ends": exon_ends,
        "strand": strand,
    }


@register_figure(FIGURES, "isoform_structure")
def isoform_structure_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "isoform_table")
    if p is None:
        raise FileNotFoundError("manifest.isoform_table required")
    cols = _load_isoforms(p)
    if cols is None:
        raise FileNotFoundError(f"unparseable isoform table: {p}")
    return isoform_structure(
        frame=cols,
        title="Isoform structure",
        out=out,
        strand_col="strand",
    )


@register_figure(FIGURES, "sashimi")
def sashimi_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "junction_table")
    if p is None:
        raise FileNotFoundError("manifest.junction_table required")
    cols = load_tsv_columns(
        p,
        {
            "junction": ("junction", "intron", "id"),
            "count": ("count", "n_reads", "support"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable junction table: {p}")
    return sashimi(frame=cols, title="Splice-junction sashimi", out=out)

File: crates/ecaa-conformance/tests/fixtures/minimal-package-real/runtime/plotting_r/stages/isoform_calling.R
# R-side isoform_calling stage: isoform_structure, sashimi.
# Mirrors lib/plotting/stages/isoform_calling.py.
#
# Plan reference: §S13.6 — long-read RNA-seq isoform figures. The
# isoform table carries comma-list cells the generic .swfc_load_tsv
# helper can't handle, so the loader here parses them inline (matches
# the Python `_load_isoforms` function).

if (!exists("swfc_register_figure")) {
  stop("source runtime/plotting_r/core.R before this stage module")
}

.iso_parse_int_list <- function(s) {
  if (is.null(s) || is.na(s) || nchar(s) == 0L) return(integer(0))
  parts <- strsplit(as.character(s), ",", fixed = TRUE)[[1]]
  parts <- trimws(parts)
  parts <- parts[nchar(parts) > 0L]
  suppressWarnings(as.integer(parts))
}

.iso_load_isoforms <- function(path) {
  df <- tryCatch(
    utils::read.delim(path, stringsAsFactors = FALSE, check.names = FALSE),
    error = function(e) NULL
  )
  if (is.null(df) || nrow(df) == 0L) return(NULL)
  cols <- tolower(colnames(df))
  t_col <- which(cols %in% c("transcript", "transcript_id", "tx"))
  s_col <- which(cols %in% c("exon_starts", "starts", "blockstarts"))
  e_col <- which(cols %in% c("exon_ends", "ends", "blockends"))
  sd_col <- which(cols %in% c("strand"))
  if (length(t_col) == 0L || length(s_col) == 0L || length(e_col) == 0L) return(NULL)
  rows <- list()
  for (i in seq_len(nrow(df))) {
    starts <- .iso_parse_int_list(df[[s_col[1]]][i])
    ends <- .iso_parse_int_list(df[[e_col[1]]][i])
    if (length(starts) == 0L || length(ends) == 0L) next
    n <- min(length(starts), length(ends))
    rows[[length(rows) + 1L]] <- data.frame(
      transcript = rep(df[[t_col[1]]][i], n),
      exon_starts = starts[seq_len(n)],
      exon_ends = ends[seq_len(n)],
      strand = if (length(sd_col)) df[[sd_col[1]]][i] else "+",
      stringsAsFactors = FALSE
    )
  }
  if (length(rows) == 0L) return(NULL)
  do.call(rbind, rows)
}

swfc_register_figure("isoform_calling", "isoform_structure", function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "isoform_table")
  if (is.null(p)) stop("manifest.isoform_table required")
  df <- .iso_load_isoforms(p)
  if (is.null(df)) stop(sprintf("unparseable isoform table: %s", p))
  swfc_isoform_structure_r(df, title = "Isoform structure",
                            strand_col = "strand")
})

swfc_register_figure("isoform_calling", "sashimi", function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "junction_table")
  if (is.null(p)) stop("manifest.junction_table required")
  df <- .swfc_load_tsv(p, list(
    junction = c("junction", "intron", "id"),
    count = c("count", "n_reads", "support")
  ))
  if (is.null(df)) stop(sprintf("unparseable junction table: %s", p))
  swfc_sashimi_r(df, title = "Splice-junction sashimi")
})
